### DISCUSSION TITLE
Google Drive: Error 401: invalid_client

### DISCUSSION THREAD

**Discussion by oatmealm** [open-webui/open-webui#8138] [Issues]
Now it tries to authenticate but seeing this error.

```
Access blocked: Authorization Error

The OAuth client was not found.
If you are a developer of this app, see [error details](https://accounts.google.com/).
Error 401: invalid_client
```

**Comment by mrrtmob** [comment]
You're encountering an "invalid_client" error with your Google Drive integration. Here's how to resolve it:

1. **Check `GOOGLE_DRIVE_CLIENT_ID`**: Ensure that this OAuth 2.0 client ID is correctly configured in the Google Cloud Console for your project.

2. **Verify `GOOGLE_DRIVE_API_KEY`**: Make sure this API key is set up for the Picker API within the same project.

3. **Enable APIs**: Confirm that both Google Drive API and Picker API are enabled in the Google Cloud Console.

4. **OAuth Consent Screen**: Ensure that the OAuth consent screen is properly configured.

**Comment by m8dhouse** [comment]
Hi - I'm trying to add this to an existing project already running oAuth Google.

I enabled the Picker and Google Drive API and generated a API key (locked to those 2 scopes, but default for the rest).
I added the Google Drive drive /auth/drive.readonly scope 
Added the variables (client ID = same as I had before for the oAuth) & generated API key. 
Login works fine, but when I try to access the Google drive option I still get:

You can’t sign in because this app sent an invalid request. You can try again later or contact the developer about this issue. [Learn more about this error](https://developers.google.com/identity/protocols/oauth2/javascript-implicit-flow#authorization-errors-origin-mismatch)
If you are a developer of this app, see [error details](https://accounts.google.com/).
Error 400: redirect_uri_mismatch


Console:
Initializing Google Drive Picker...
google-drive-picker.ts:112 Getting auth token...
google-drive-picker.ts:208 Google Drive Picker error: Error: Popup window closed
    at Ls.error_callback (google-drive-picker.ts:89:13)
    at _.pj.D (client:231:350)
    at pg (client:197:193)
    at _.l.dispatchEvent (client:195:348)
    at _.kj.F (client:222:551)
(anonymous) @ google-drive-picker.ts:208
await in (anonymous)
td @ google-drive-picker.ts:108
It @ MessageInput.svelte:642
R @ InputMenu.svelte:161
(anonymous) @ lifecycle.js:105
(anonymous) @ lifecycle.js:104
(anonymous) @ events.js:8
C @ event.js:43
(anonymous) @ event.js:48Understand this errorAI
MessageInput.svelte:652 Google Drive Error: Error: Popup window closed

Any ideas? Is there another redirect callback URL needed?

**Comment by richtong** [comment]
I"m such a noob on this (I've done this a zillion times but never write it down, but how do you get the Client ID, what is the callback address, normally for webapps there is a callback address. So I divined that you have to:

1. Enable Google Drive in the Admin Settings > Settings > Documents > Google Drive
2. You create two environment variables. Enable both the Picker and the Drive APIs
3. Create an API Key Credential limited to these two APIs. Set the key to GOOGLE_DRIVE_API_KEY
4. Go back to credentials and choose Create Credential > Oauth key
5. Set the client url to "http://localhost:8080", get the Client_ID strng and create the GOOGLE_DRIVE_CLIENT_ID
6. Crank up openwebui and go to Admin Settings > Settings > Document and Google Drive Enable
7. Now go to Chat and pick on the + and Google Drive appears, click on it
8. YOu get the Oauth screen and I login fine
9. But then I get "The API Developer Key is invalid"

How can I debug this, I can't find anywhere where the two environment variables are stored 

<img width="1624" alt="image" src="https://github.com/user-attachments/assets/f8a0b5bf-20b2-4a2b-8eaa-a7de754d1f77" />

**Comment by shashanknimje** [comment]
I was running into similar issues.  Here's what worked for me.

1. Create an **OAuth 2.0 client** and configure both the **Authorized JavaScript origins** & **Authorized redirect URI** to be the URL (include the port if any) you use to access your Open-WebUI instance.
2. Make a note of the Client ID associated with that OAuth client.
3. Make sure that you enable both **Google Drive API** and **Google Picker API** for your project.
4. Also set your app (project) as **Testing** and add your Google Drive email to the **User List**
5. Set the permission scope to include everything those APIs have to offer. And because the app would be in **Testing** mode, no verification is required by Google to allow the app from accessing the data of the limited test users.
6. Go to the **Google Picker API** page, and click on the **create credentials** button.
7. Create an API key and under **Application restrictions** and choose **Websites**. Then add your Open-WebUI instance's URL, same as the **Authorized JavaScript origins** and **Authorized redirect URIs** settings in the step 1.
8. Set up **API restrictions** on the API Key to only have access to **Google Drive API** & **Google Picker API**
9. Set up the environment variable, **GOOGLE_DRIVE_CLIENT_ID** to the Client ID of the OAuth client from step 2.
10. Set up the environment variable **GOOGLE_DRIVE_API_KEY** to the API Key value setup up in step 7 (**NOT** the OAuth client secret from step 2).
11. Set up the **GOOGLE_REDIRECT_URI** to my Open-WebUI instance's URL (include the port, if any).
12. Then relaunch your Open-WebUI instance with those three environment variables.
13. After that, make sure Google Drive was enabled under **Admin Panel < Settings < Documents < Google Drive**

After following those steps, I was able to integrate Google Drive with Open-WebUI successfully.

**Reply by glenviewjeff** [reply]
@shashanknimje thank you, this fixed my problems.

**Comment by richtong** [comment]
Yes I didn't do the testing thing so it does require authentication and the creation of the oauth application so more involved but also allow logins to different google accounts if you have (eg work vs personal).

Sent via Superhuman iOS ( ***@***.*** )

**Reply by shashanknimje** [reply]
I believe the "Testing" option allows up to 100 users (email accounts). It might be worth giving it a try.

**Comment by hdnh2006** [comment]
I have made a tutorial about how to integrate Google Drive into Open WebUI

- Here you have the free medium article: https://henrynavarro.org/how-to-integrate-google-drive-with-open-webui-e525c8f3f82e?sk=a9f48cd1590e3e8dca052f7fe9f12765
- And for those spanish speakers 🇪🇸, you have a tutorial in my channel: https://youtu.be/cGBjAned72A?si=99TRUKgHZQXWcNs2
